core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala

```diff
@@ -0,0 +1,258 @@
+    endpointRef: NettyRpcEndpointRef,
+    override val endpoint: ThreadSafeRpcEndpoint) extends Inbox(endpointRef, endpoint) {
+
+  private val _endpoint = endpoint.asInstanceOf[ThreadSafeRpcEndpoint]
+
+  // protected by "this"
+  private var stopped = false
+
+  // OnStart should be the first message to process
+  messages.add(OnStart)
+
+  override def process(dispatcher: Dispatcher): Boolean = {
+    var exit = false
+    var message = messages.poll()
+    while (message != null) {
+      safelyCall(endpoint) {
+        message match {
+          case ContentMessage(_sender, content, needReply, context) =>
+            val pf: PartialFunction[Any, Unit] =
+              if (needReply) {
+                endpoint.receiveAndReply(context)
+              } else {
+                endpoint.receive
+              }
+            try {
+              pf.applyOrElse[Any, Unit](content, { msg =>
+                throw new SparkException(s"Unmatched message $message from ${_sender}")
+              })
+              if (!needReply) {
+                context.finish()
+              }
+            } catch {
+              case NonFatal(e) =>
+                if (needReply) {
+                  // If the sender asks a reply, we should send the error back to the sender
+                  context.sendFailure(e)
+                } else {
+                  context.finish()
+                  throw e
+                }
+            }
+
+          case OnStart => {
+            _endpoint.asInstanceOf[ThreadSafeRpcEndpoint].onStart()
+          }
+          case OnStop =>
+            dispatcher.unregisterRpcEndpoint(endpointRef.name)
+            _endpoint.onStop()
+            assert(isEmpty, "OnStop should be the last message")
+            exit = true
+          case Associated(remoteAddress) =>
+            endpoint.onConnected(remoteAddress)
+          case Disassociated(remoteAddress) =>
+            endpoint.onDisconnected(remoteAddress)
+          case AssociationError(cause, remoteAddress) =>
+            endpoint.onNetworkError(cause, remoteAddress)
+        }
+      }
+      message = messages.poll()
+    }
+    exit
+  }
+
+  override def post(message: InboxMessage): Unit = {
+    val dropped =
+      synchronized {
+        if (stopped) {
+          // We already put "OnStop" into "messages", so we should drop further messages
+          true
+        } else {
+          messages.add(message)
+          false
+        }
+      }
+    if (dropped) {
+      onDrop()
+    }
+  }
+
+  override def stop(): Unit = synchronized {
+    // The following codes should be in `synchronized` so that we can make sure "OnStop" is the last
+    // message
+    if (!stopped) {
+      stopped = true
+      messages.add(OnStop)
+    }
+  }
+}
+
+/**
+ * A inbox that stores messages for an [[RpcEndpoint]] and posts messages to it concurrently.
+ * @param endpointRef
+ * @param endpoint
+ */
+private[netty] class ConcurrentInbox(
```

Comment:
So I'm reading this code and I'm starting to think that you don't need two inboxes. The thread-safety guarantees for endpoints are not a result of this inbox being "thread-safe" in some way, but because the message dispatcher makes sure only one thread is dispatching messages for a `ThreadSafeEndpoint` at a time.

I think you could get away with a single inbox implementation for both cases, and leave the "single thread dispatching" semantics to the `MessageLoop` class. For example, you could have an `AtomicBoolean` for each thread-safe endpoint that could be used as cheap locks.